File: CPP/ByRound/1729/1729_F_Kirei_and_the_Linear_Function.cpp
// Problem: CF 1729 F - Kirei and the Linear Function
// https://codeforces.com/contest/1729/problem/F

/*
Purpose: Solve the problem of finding two substrings of length w such that a specific modular arithmetic condition holds.
Approach:
- Precompute prefix sums for digit strings to quickly calculate substring values modulo 9.
- Group all substrings of length w by their value modulo 9.
- For each query, determine valid pairs of substrings based on the formula (v(L1) * v(l,r) + v(L2)) % 9 == k.

Time Complexity: O(n * w + m * 9^2) 
Space Complexity: O(n + 9 * w) 
*/
#include <algorithm>
#include <array>
#include <iostream>
#include <string>
#include <vector>

using namespace std;
#define ll long long
int main() {
  int Tit = 1;
  cin >> Tit;
  while (Tit--) {
    string s;
    cin >> s;
    int n = s.length();
    int w, m;
    cin >> w >> m;
    // Precompute prefix sums for efficient substring sum calculation
    vector<int> f(n + 1);
    for (int i = 0; i < n; i++)
      f[i + 1] = f[i] + s[i] - '0';
    
    // Group substrings of length w by their sum mod 9
    vector<int> a[9];
    for (int i = 0; i + w <= n; i++)
      a[(f[i + w] - f[i]) % 9].push_back(i);
    
    while (m--) {
      // Initialize answer with large values
      array<int, 2> ans{n, n};
      int l, r, k;
      cin >> l >> r >> k;
      // Get the value of query substring modulo 9
      int x = (f[r] - f[l - 1]) % 9;
      
      // Check all possible pairs of groups for valid combinations
      for (int u = 0; u < 9; u++) {
        // Case 1: Two substrings from the same group
        if (a[u].size() > 1 && (u * x + u) % 9 == k)
          ans = min(ans, {a[u][0], a[u][1]});
        
        // Case 2: Two substrings from different groups
        for (int v = 0; v < 9; v++) {
          if (u != v && a[u].size() && a[v].size() && (u * x + v) % 9 == k)
            ans = min(ans, {a[u][0], a[v][0]});
        }
      }
      
      // Output result or -1 -1 if not found
      if (ans[0] == n)
        cout << "-1 -1\n";
      else
        cout << ans[0] + 1 << ' ' << ans[1] + 1 << endl;
    }
  }
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/F/1729_F_Kirei_and_the_Linear_Function.py
# Problem: CF 1729 F - Kirei and the Linear Function
# https://codeforces.com/contest/1729/problem/F

"""
Algorithm: Dynamic Programming with Modular Arithmetic
Approach:
- Precompute prefix sums modulo 9 for efficient substring modulo calculation
- For each possible remainder (0-8), store the first two indices of substrings with that remainder
- For each query, calculate the required remainder of the query substring
- Find valid pairs of substrings that satisfy the equation: (L1 * v_query + L2) % 9 == k
- Use stored indices to quickly find valid pairs

Time Complexity: O(n + m)
Space Complexity: O(n)

Where n is the length of string s and m is the number of queries.
"""

for _ in range(int(input())):
    s = list(map(int, input()))
    w, q = map(int, input().split())
    n = len(s)
    # Precompute prefix sums modulo 9 for quick substring modulo calculation
    pre = [0]
    for x in s:
        pre += [(x + pre[-1]) % 9]

    # Store first two indices for each possible remainder (0-8)
    ind = [[] for _ in range(9)]
    for i in range(n - w + 1):
        x = (pre[i + w] - pre[i] + 9) % 9  # Calculate substring remainder
        if len(ind[x]) == 2:
            continue  # Only keep first two indices for efficiency
        ind[x] += [i + 1]  # Store 1-based index

    for _ in range(q):
        l, r, k = map(int, input().split())
        # Calculate remainder of query substring
        x = (pre[r] - pre[l - 1] + 9) % 9
        ans = [n, n]  # Initialize with large values

        # Check all possible remainders for valid pairs
        for i in range(9):
            if not ind[i]:
                continue
            # Calculate required remainder for second substring
            j = (k - i * x + 9) % 9
            if i == j:
                # Same remainder: need two different substrings
                if len(ind[i]) != 2:
                    continue
                ans = min(ans, ind[i])
            else:
                # Different remainders: find first occurrence of both
                if not ind[j]:
                    continue
                ans = min(ans, [ind[i][0], ind[j][0]])

        if ans == [n, n]:
            print(-1, -1)
        else:
            print(*ans)


# https://github.com/VaHiX/CodeForces/